x-pack/auditbeat/module/system/user/user.go

```diff
@@ -0,0 +1,431 @@
+type MetricSet struct {
+	mb.BaseMetricSet
+	config     Config
+	log        *logp.Logger
+	cache      *cache.Cache
+	bucket     datastore.Bucket
+	lastState  time.Time
+	lastChange time.Time
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+	if runtime.GOOS != "linux" {
+		return nil, fmt.Errorf("the %v/%v dataset is only supported on Linux", moduleName, metricsetName)
+	}
+
+	config := defaultConfig
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
+		cache:         cache.New(),
+		bucket:        bucket,
+	}
+
+	// Load from disk: Time when state was last sent
+	err = bucket.Load(bucketKeyStateTimestamp, func(blob []byte) error {
+		if len(blob) > 0 {
+			return ms.lastState.UnmarshalBinary(blob)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if !ms.lastState.IsZero() {
+		ms.log.Debugf("Last state was sent at %v. Next state update by %v.", ms.lastState, ms.lastState.Add(ms.config.effectiveStatePeriod()))
+	} else {
+		ms.log.Debug("No state timestamp found")
+	}
+
+	// Load from disk: Users
+	users, err := ms.restoreUsersFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore users from disk")
+	}
+	ms.log.Debugf("Restored %d users from disk", len(users))
+
+	ms.cache.DiffAndUpdateCache(convertToCacheable(users))
+
+	return ms, nil
+}
+
+// Close cleans up the MetricSet when it finishes.
+func (ms *MetricSet) Close() error {
+	if ms.bucket != nil {
+		return ms.bucket.Close()
+	}
+	return nil
+}
+
+// Fetch collects the user information. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	needsStateUpdate := time.Since(ms.lastState) > ms.config.effectiveStatePeriod()
+	if needsStateUpdate || ms.cache.IsEmpty() {
+		ms.log.Debugf("State update needed (needsStateUpdate=%v, cache.IsEmpty()=%v)", needsStateUpdate, ms.cache.IsEmpty())
+		err := ms.reportState(report)
+		if err != nil {
+			ms.log.Error(err)
+			report.Error(err)
+		}
+		ms.log.Debugf("Next state update by %v", ms.lastState.Add(ms.config.effectiveStatePeriod()))
+	}
+
+	err := ms.reportChanges(report)
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+	}
+}
+
+// reportState reports all existing users on the system.
+func (ms *MetricSet) reportState(report mb.ReporterV2) error {
+	ms.lastState = time.Now()
+
+	users, err := GetUsers()
+	if err != nil {
+		return errors.Wrap(err, "failed to get users")
+	}
+	ms.log.Debugf("Found %v users", len(users))
```

Comment:
Total user count could also be a good shorthand to save on a checkin / snapshot.

Reply:
You mean send it in every user document? I guess we could, but maybe also here we could wait until we have an clear need (i.e. the UI would benefit from it).

Reply:
Hmmm, I may have assumed something from our original discussions. In my mind, there's one document being sent with overall details about a "checkin", and multiple associated documents, one per user, one per process, etc. So I would have saved that user count (and group count for that matter) on that one checkin document.

But thinking back about my review, I don't remember seeing such a generic checkin document, you're right. So I guess disregard for now. But it may be useful metadata, if we ever decide we need a single "checkin" document, to tie all other documents around it in the future.